### PR TITLE
Introduce `ExplicitArgumentEnumeration` check

### PR DESCRIPTION
**Pull request by Stephan202**
Suggested commit message:
```
Introduce `ExplicitArgumentEnumeration` check (#985)
```

I found this on an old branch from early last year. Extended it a bit and now putting it up for review. Perhaps we can find a better name. Likely this code should also be tested against the internal code base before landing. But it's certainly ready for review.

### CODE REVIEW

File: error-prone-contrib/src/main/java/tech/picnic/errorprone/bugpatterns/ExplicitArgumentEnumeration.java [hunk 1]
```diff
@@ -0,0 +1,200 @@
...
+            : SuggestedFix.builder()
+                .merge(SuggestedFixes.renameMethodInvocation(tree, alternative, state))
+                .merge(fix)
+                .build());
+  }
+
+  // XXX: Once we target JDK 14+, drop this method in favour of `Symbol#isPublic()`.
+  private static boolean isPublic(Symbol symbol) {
+    return symbol.getModifiers().contains(Modifier.PUBLIC);
+  }
```
Comment (Stephan202) [inline error-prone-contrib/src/main/java/tech/picnic/errorprone/bugpatterns/ExplicitArgumentEnumeration.java:199]: Also present in #972, but a separate utility method doesn't seem worth it. ([Or we can just drop JDK 11 support](https://github.com/PicnicSupermarket/error-prone-support/pull/603#issuecomment-1902727809) :upside_down_face:.)

**Review by Stephan202 — COMMENTED** [review]
(I see there a more mutants to be handled; will review in the ~coming days.)

File: error-prone-contrib/src/main/java/tech/picnic/errorprone/bugpatterns/ExplicitArgumentEnumeration.java [hunk 2]
```diff
@@ -0,0 +1,200 @@
...
+        .map(fix -> describeMatch(tree, fix))
+        .orElse(Description.NO_MATCH);
+  }
+
+  private static boolean isUnaryIterableAcceptingMethod(MethodSymbol method, VisitorState state) {
+    List<VarSymbol> params = method.params();
+    return !method.isVarArgs()
+        && params.size() == 1
+        && ASTHelpers.isSubtype(params.get(0).type, state.getSymtab().iterableType, state);
+  }
```
Comment (Stephan202) [inline error-prone-contrib/src/main/java/tech/picnic/errorprone/bugpatterns/ExplicitArgumentEnumeration.java:132]: Pitest flags that the `params.size() == 1` can be mutated away. This is true thanks to the initial `tree.getArguments().size() != 1` check, but it doesn't seem right to drop this condition; it makes the code less clear, IMHO.

**Review by Stephan202 — COMMENTED** [review]
Rebased and added a commit with some improvements. I think this is now properly ready for review.

**Comment by rickie** [comment]
```
Error:  Failed to execute goal org.apache.maven.plugins:maven-compiler-plugin:3.13.0:compile (error-prone-compile) on project checkstyle: Fatal error compiling: The default --should-stop=ifError policy (INIT) is not supported by Error Prone, pass --should-stop=ifError=FLOW instead -> [Help 1]
```
From the IT ☝🏻.


Not checking the error in depth right now, but curious 🤔.

**Comment by rickie** [comment]
And this....

```
Caused by: org.eclipse.aether.transfer.ArtifactTransferException: com.github.PicnicSupermarket.error-prone:error_prone_core:jar:v2.36.0-picnic-1 failed to transfer from https://jitpack.io/ during a previous attempt. This failure was cached in the local repository and resolution is not reattempted until the update interval of jitpack.io has elapsed or updates are forced. Original error: Could not transfer artifact com.github.PicnicSupermarket.error-prone:error_prone_core:jar:v2.36.0-picnic-1 from/to jitpack.io (https://jitpack.io/): Checksum validation failed, no checksums available
```

**Comment by Stephan202** [comment]
> ```
> Error:  Failed to execute goal org.apache.maven.plugins:maven-compiler-plugin:3.13.0:compile (error-prone-compile) on project checkstyle: Fatal error compiling: The default --should-stop=ifError policy (INIT) is not supported by Error Prone, pass --should-stop=ifError=FLOW instead -> [Help 1]
> ```
> 
> From the IT ☝🏻.
> 
> Not checking the error in depth right now, but curious 🤔.

I filed #1455 for this.

**Comment by Stephan202** [comment]
> And this....
> 
> > Caused by: org.eclipse.aether.transfer.ArtifactTransferException: com.github.PicnicSupermarket.error-prone:error_prone_core:jar:v2.36.0-picnic-1 failed to transfer from [jitpack.io](https://jitpack.io/) during a previous attempt. This failure was cached in the local repository and resolution is not reattempted until the update interval of jitpack.io has elapsed or updates are forced. Original error: Could not transfer artifact com.github.PicnicSupermarket.error-prone:error_prone_core:jar:v2.36.0-picnic-1 from/to jitpack.io ([jitpack.io](https://jitpack.io/)): Checksum validation failed, no checksums available

Same also happened for a `master` build, but a restart fixed it. The new PR mentioned above also has a passing build.

**Comment by rickie** [comment]
Oops, sorry, don't know what happened there 😓.

File: error-prone-contrib/src/main/java/tech/picnic/errorprone/bugpatterns/ExplicitArgumentEnumeration.java [hunk 3]
```diff
@@ -0,0 +1,200 @@
...
+
+  private static Optional<SuggestedFix> trySuggestCallingCustomAlternative(
+      MethodInvocationTree tree, MethodInvocationTree argument, VisitorState state) {
+    return ALTERNATIVE_METHODS.rowMap().entrySet().stream()
+        .filter(e -> e.getKey().matches(tree, state))
+        .findFirst()
+        .flatMap(e -> trySuggestCallingCustomAlternative(tree, argument, state, e.getValue()));
+  }
+
+  private static Optional<SuggestedFix> trySuggestCallingCustomAlternative(
```
Comment (rickie) [inline error-prone-contrib/src/main/java/tech/picnic/errorprone/bugpatterns/ExplicitArgumentEnumeration.java:180]: Probably the only exception where `VisitorState` is not the last parameter :).

File: error-prone-contrib/src/main/java/tech/picnic/errorprone/bugpatterns/ExplicitArgumentEnumeration.java [hunk 4]
```diff
@@ -0,0 +1,200 @@
...
+            alternatives.get(ASTHelpers.getSymbol(tree).getSimpleName().toString()))
+        .map(
+            replacement ->
+                SuggestedFix.builder()
+                    .merge(SuggestedFixes.renameMethodInvocation(tree, replacement, state))
+                    .merge(SourceCode.unwrapMethodInvocation(argument, state))
+                    .build());
+  }
+
+  private static boolean isAtLeastAsVisible(Element symbol, Element reference) {
```
Comment (rickie) [inline error-prone-contrib/src/main/java/tech/picnic/errorprone/bugpatterns/ExplicitArgumentEnumeration.java:195]: Was thinking that the name `modifier` in this method name would be nice as addition, but also  fine with leaving it as is.

**Review by rickie — APPROVED** [review]
Only one small typo and a suggestion :). Nice check @Stephan202 . Sorry for the slow review 🐌.

File: error-prone-contrib/src/main/java/tech/picnic/errorprone/bugpatterns/ExplicitArgumentEnumeration.java (hunk 3 above)
Comment (Stephan202) [reply error-prone-contrib/src/main/java/tech/picnic/errorprone/bugpatterns/ExplicitArgumentEnumeration.java:180]: Yeah, that's to appease the "consistent overloads" rule. The alternative is to find another method name.

File: error-prone-contrib/src/main/java/tech/picnic/errorprone/bugpatterns/ExplicitArgumentEnumeration.java (hunk 4 above)
Comment (Stephan202) [reply error-prone-contrib/src/main/java/tech/picnic/errorprone/bugpatterns/ExplicitArgumentEnumeration.java:195]: The modifier is a syntactic means of influencing visibility, but for the purpose of this method that doesn't seem like a relevant detail.

File: error-prone-contrib/src/test/java/tech/picnic/errorprone/bugpatterns/ExplicitArgumentEnumerationTest.java [hunk 5]
```diff
@@ -0,0 +1,161 @@
...
+            "  void m() {",
+            "    ImmutableList.builder().addAll(ImmutableList.of()).build();",
+            "",
+            "    assertThat(ImmutableList.of()).containsAnyElementsOf(ImmutableMultiset.of());",
+            "    assertThat(ImmutableList.of()).containsAll(ImmutableSet.of());",
+            "    assertThat(ImmutableList.of()).containsExactlyElementsOf(List.of());",
+            "    assertThat(ImmutableList.of()).containsExactlyInAnyOrderElementsOf(Set.of());",
+            "    assertThat(ImmutableList.of()).containsSequence(Arrays.asList());",
+            "    assertThat(ImmutableList.of()).containsSubsequence(ImmutableList.of(1));",
+            "    assertThat(ImmutableList.of()).doesNotContainAnyElementsOf(ImmutableMultiset.of(1));",
```
[the remark this answers is not in the document]
Comment (Stephan202) [reply error-prone-contrib/src/test/java/tech/picnic/errorprone/bugpatterns/ExplicitArgumentEnumerationTest.java]: Hmm, the number is per statement here, just like in the `identification` test above. Will tweak here, but leave the rest as-is.

File: pom.xml [hunk 6]
```diff
@@ -454,6 +454,11 @@
                 <artifactId>value-annotations</artifactId>
                 <version>2.10.1</version>
             </dependency>
+            <dependency>
+                <groupId>org.jooq</groupId>
+                <artifactId>jooq</artifactId>
+                <version>3.16.23</version>
```
[the remark this answers is not in the document]
Comment (Stephan202) [reply pom.xml]: Well, maybe the age of this PR has something to do with that? :upside_down_face:

**Review by Stephan202 — COMMENTED** [review]
Rebased and added a commit. @mohamedsamehsalah WDYT? :)

**Review by mohamedsamehsalah — APPROVED** [review]
Clean 🧼 

(Sorry this slipped form my notifications 🙏 )

**Comment by Stephan202** [comment]
Changed the scope of the new dependency. @rickie: also here, feel free to merge if you agree with the final changes.

File: pom.xml (hunk 6 above)
[the remark this answers is not in the document]
Comment (rickie) [reply pom.xml]: Haha, my bad. I just remembered in JPLA that JOOQ usually had some edge cases. Didn't check on which version we are internally but was thinking maybe there was something related to that .

File: error-prone-contrib/src/main/java/tech/picnic/errorprone/bugpatterns/ExplicitArgumentEnumeration.java (hunk 3 above)
Comment (rickie) [reply error-prone-contrib/src/main/java/tech/picnic/errorprone/bugpatterns/ExplicitArgumentEnumeration.java:180]: No it's fine, just found it funny that indeed the overload rule caused the first exception :).

**Pull request merged**